Description for extensions (#82)
owncloud/docs-ocis: modules/ROOT/pages/deployment/extensions/extensions.adoc

These statements are true when using the runtime aswell when running extensions independent of the runtime.

image::deployment/extensions/concurrence.drawio.svg[width=600]

== Restarting a Runtime Extension

You can stop or restart a runtime extension, but some things are important to understand. These are:

* You can only kill and start a runtime extension.
* You can not instantinate a runtime extension, an error will occur.
* Special rules apply to restarted runtime extensions with respect to their configuration.
* All runtime extensions share the same PID of `ocis server`.
* Runtime extensions share the same PID while non runtime extensions have their own PID.

Note that in the examples described, only the webdav extension is used, but the process is valid for all extensions available.

The description in the example may sound complex, but isnt. The ruleset is straight forward. The images below helps to get it in one view.

Example 1::
a. You start the oCIS runtime with `ocis server`
+
--
. The configuration file `/etc/ocis/ocis.yaml` is read.
. The configuration file `/etc/ocis/webdav.yaml` is read AND applied.

> . The environment variables are taken, applied and may overwrite already existing configuration elements. See section xref:deployment/general/general-info.adoc#configuration-of-ocis[Configuration of oCIS] about the order it got applied (config arithmetics).

Reviewer: ```suggestion
. The environment variables are read and applied. These can overwrite already existing configuration elements. See section xref:deployment/general/general-info.adoc#configuration-of-ocis[Configuration of oCIS] about the order in which different configurations get applied (config arithmetics).
```

--
+
All extensions have started, share the same PID and have a defined config. Typing `ocis list` shows all active runtime extensions.
+

> b. By command, the webdav extension is killed with `ocis kill webdav`.

Reviewer: ```suggestion
b. On the command line, the WebDAV extension is killed with `ocis kill webdav`.
```

+
Listing the active runtime extensions with `ocis list` shows that webdav is no longer running.
+
c. The webdav extension gets restarted with `ocis run webdav`.
+

> Note the envoronment variable `webdav=setting2` +

Reviewer: ```suggestion
Note the environment variable `webdav=setting2` +
```

> and the command line option `--config-file=webdav.yaml`

Reviewer: ```suggestion
and the command line option `--config-file=webdav.yaml`.
```

+
[arabic,start=4]

> . *Following config rules apply to configure the webdav runtime extension:*

Reviewer: ```suggestion
. *The following config rules apply when configuring the WebDAV runtime extension:*
```

+
--
... The _oics configuration_ and the _environment variables_ are taken from the _runtime_ (1 and 3).
... The configuration file for the extension from the runtime (2) is NOT used.
... Environment variables from the command `ocis run` (4) are NOT used. Generally, you can omit them.

> ... The config file defined as command line option (4) is taken and the config arithmetics is applied for a final config for the webdav extension.

Reviewer: ```suggestion
... The config file defined as command line option (4) is taken and the config arithmetic is applied for a final config for the WebDAV extension.
```

--
+

> When monitoring running PID's, the webdav extension does not show up as it is part of the ocis server PID.

Reviewer: ```suggestion
When monitoring running PIDs, the WebDAV extension does not show up as it is part of the ocis server process and shares its PID.
```

> WARNING: If you have started ocis server with a config file for an extension, you must define it in the command line option when restarting. If you omit it when restarting, the extension will be configured partially leading to unexpected behaviour.

Reviewer: ```suggestion
WARNING: If you have started ocis server with a config file for an extension, you must define it in the command line option when restarting. If you omit it, the extension will be configured partially leading to unexpected behavior.
```

> This image gives you a graphical represeantation about the ruleset described above.

Reviewer: ```suggestion
This image gives you a graphical representation of the rule set described above.
```

image::deployment/extensions/runtime_c1.drawio.svg[width=600]

Example 2::

> The same rules apply like in the example using one server, when using another server from where the kill and run commands to manage an extensions are set. Note that the config file origins from server 2.

Reviewer: ```suggestion
The same rules as in the example with one server apply when using an additional server from which to issue kill and run commands to manage extensions. Note that the config file originates from server 2.
```
